content/en/blog/_posts/2022-09-14-pod-has-network-condition.md

```diff
@@ -0,0 +1,115 @@
+---
+layout: blog
+title: 'Kubernetes 1.25: PodHasNetwork condition for pods'
+date: 2022-09-14
+slug: pod-has-network-condition
+---
+
+**Author:**
+Deep Debroy (Apple)
+
+Kubernetes 1.25 introduces Alpha support for a new kubelet-managed pod condition
+in the status field of a pod: `PodHasNetwork`. The kubelet will use the
+`PodHasNetwork` condition to accurately surface the initialization state of a
+pod from the perspective of pod sandbox creation and network configuration by a
+CRI plugin in coordination with CNI plugins. The kubelet starts to pull
+container images and start individual containers (including init containers)
+after the status of the PodHasNetwork condition is set to `True`. Metrics
+collection services that report latency of pod initialization from a cluster
+infrastructural perspective (i.e. agnostic of per container characteristics like
+image size or payload) can utilize the `PodHasNetwork` condition to accurately
+generate SLIs. Certain operators or controllers that manage underlying pods may
+utilize the PodHasNetwork condition to optimize the set of actions performed
+when pods repeatedly fail to come up.
+
+### How is this different from the existing Initialized condition reported for pods?
+
+The kubelet sets the status of the existing `Initialized` condition reported in
+the status field of a pod depending on the presence of init containers in a pod.
+
+If a pod specifies init containers, the status of the `Initialized` condition in
+the pod status will not be set to `True` until all init containers for the pod
+have succeeded. In large multi-tenant clusters, init containers are owned and
+configured by tenant workloads. Thus, cluster-wide, infrastructural SLIs cannot
+depend on the `Initialized` condition of pods as init containers may have user
+errors (payload crashing, invalid image, etc) and the number of init containers
+configured in a pod may vary.
+
+If a pod does not specify init containers, the status of the `Initialized`
+condition in the pod status is set to `True` very early in the lifecycle of the
+pod. This occurs before the kubelet initiates any pod runtime sandbox creation
+and network configuration steps. As a result, a pod without init containers will
+report the status of the `Initialized` condition as `True` even if the container
+runtime is not able to successfully initialize the pod sandbox environment.
+
+Relative to either situation above, the `PodHasNetwork` condition surfaces more
+accurate data around when the pod runtime sandbox was initialized with
+networking configured so that the kubelet can proceed to launch user-configured
+containers (including init containers) in the pod.
+
```

Comment:
if you think this needs more clarification in the general context for most pod lifecycles ..  feel free to leave it out.. we can discuss further

Reply:
Is this special behavior strictly something that happens in the cloud? (could an on-prem cluster do something similar?)

**If** we're adding text here, we could make it clear that `PodHasNetwork` means that for each configured address family, the kubelet sees that [_the one IP address that a conformant network plugin should set up for that address family_] is configured and up.

Would need better wording than mine, mind!

Reply:
FYI: CRI (or wider - cloud providers, as suggested by @mikebrow) can do anything it/they want at any point of pod lifecycle. it can call a cni plugin, or multiple of them, or whatever it wants. but that's out of the scope of the description of "kubernetes contract".
i didn't yet red the mentioned KEP so i'm not touching the topic how the data should be explained, but looking on @mikebrow description i would expect that `PodHasNetwork` will be only describing status of the default network connectivity, ignoring all "whathever will cri call/setup" things which are not the default access to kubeapi network things (like e.g. what is hold in extensions under the https://github.com/k8snetworkplumbingwg/ meaning, or more precisely under the https://docs.google.com/document/d/1Ny03h6IDVy_e_vmElOqR7UdTPAG_RNydhVE1Kx54kFQ/edit#heading=h.hylsbqoj5fxd extension)

Reply:
some more discussion/clarifications here: https://kubernetes.slack.com/archives/C09QYUH5W/p1661970934172379?thread_ts=1661969610.085099&cid=C09QYUH5W

Reply:
Thanks for the inputs and clarifications about this, @mikebrow, @sftim and @jellonek. I will summarize and call out the above point.

Reply:
Clarified this. For reference, https://github.com/maiqueb/multus-dynamic-networks-controller appears to be a concrete implementation of what was being referred to above.